File: src/di_construction_demo.cpp
/*  _______________________________________________________________________

    Dakota: Explore and predict with confidence.
    Copyright 2014-2025
    National Technology & Engineering Solutions of Sandia, LLC (NTESS).
    This software is distributed under the GNU Lesser General Public License.
    For more information, see the README file in the top Dakota directory.
    _______________________________________________________________________ */

#include "DakotaVariables.hpp"
#include "DakotaResponse.hpp"
#include "Study.hpp"
#include "StudyConfig.hpp"
#include "NonDLHSSampling.hpp"
#include "SimulationModel.hpp"

#include <iostream>
#include <memory>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

int main()
{
  using namespace Dakota;

  const json method_json = {
    {"sample_type", {{"lhs", true}}},
    {"samples", 10},
    {"seed", 1234}
  };

  const json variables_json = {
    {"active", {{"all", true}}},
    {"uniform_uncertain", {
      {"count", 2},
      {"descriptors", {"x1", "x2"}},
      {"lower_bounds", {0.0, 0.0}},
      {"upper_bounds", {1.0, 1.0}}
    }}
  };

  const json responses_json = {
    {"response_type", {{"response_functions", {{"count", 1}}}}},
    {"descriptors", {"f"}},
    {"gradient_type", {{"no_gradients", true}}},
    {"hessian_type", {{"no_hessians", true}}}
  };

  const json interface_json = {
    {"analysis_drivers", {
      {"drivers", {"text_book"}},
      {"interface_type",
        {
          {
            "fork", {
              {"parameters_file", "params.in"},
              {"results_file", "results.out"},
              {"file_save", true}
            }
          }
        }
      },
      {"deactivate", {"restart_file", true}},

    }}
  };

  const json model_json = json::object();

  StudyConfig config;
  config.output.precision = 12;
  config.output.outputFile = "di_construction_demo.out";
  config.output.errorFile = "di_construction_demo.err";
  config.output.writeRestart = "di_construction_demo.rst";
  config.output.resultsOutput = true;
  config.output.resultsOutputFile = "di_construction_demo_results";

  // Study owns the library-mode services used by DI-constructed components.
  Study study(config);

  std::cout << "Constructing DI study components...\n";
  std::cout << "Configured output precision: "
            << study.output_manager()->write_precision() << '\n';
  Variables variables = study.variables(variables_json);
  Response response = study.responses(responses_json, variables);
  auto interface = study.interface(interface_json);
  auto model = study.model().simulation(
    model_json, variables, interface, response);
  auto sampling = study.method().sampling(method_json, model);

  std::cout << "Running sampling study...\n";
  study.run(sampling);

  const auto& responses = sampling->all_responses();
  std::cout << "Completed DI study.\n";
  std::cout << "Samples evaluated: " << responses.size() << '\n';
  if (!responses.empty()) {
    const auto& first_response = responses.begin()->second;
    if (first_response.num_functions() > 0)
      std::cout << "First response value: "
                << first_response.function_value(0) << '\n';
  }

  return 0;
}

File: python/dakota/study/di_construction_demo.py
"""DI construction demo using the :mod:`dakota.study` Python bindings.

This mirrors ``src/di_construction_demo.cpp``: it assembles a small sampling
study from keyword-wrapped JSON-shaped dictionaries, constructs Dakota
components through a ``Study`` context, and runs the resulting method.
"""

from __future__ import annotations

try:
    from . import Response, Study, StudyConfig, Variables
except ImportError:
    # Allow direct execution by path during development, e.g.
    # ``python python/dakota/study/di_construction_demo.py``.
    from dakota.study import Response, Study, StudyConfig, Variables


METHOD = {
    "sampling": {
        "sample_type": {"lhs": True},
        "samples": 10,
        "seed": 1234,
    }
}

VARIABLES = {
    "active": {"all": True},
    "uniform_uncertain": {
        "count": 2,
        "descriptors": ["x1", "x2"],
        "lower_bounds": [0.0, 0.0],
        "upper_bounds": [1.0, 1.0],
    },
}

RESPONSES = {
    "response_type": {"response_functions": {"count": 1}},
    "descriptors": ["f"],
    "gradient_type": {"no_gradients": True},
    "hessian_type": {"no_hessians": True},
}

INTERFACE = {
    "analysis_drivers": {
        "drivers": ["text_book"],
        "interface_type": {
            "fork": {
                "parameters_file": "params.in",
                "results_file": "results.out",
                "file_save": True,
            }
        },
        "deactivate": {"restart_file": True},
    }
}

MODEL = {}


def main() -> None:
    config = StudyConfig()
    config.output.precision = 12
    config.output.output_file = "di_construction_demo.out"
    config.output.error_file = "di_construction_demo.err"
    config.output.write_restart = "di_construction_demo.rst"
    config.output.results_output = True
    config.output.results_output_file = "di_construction_demo_results"

    study = Study(config)

    print("Constructing DI study components...")
    print(f"Configured output precision: {config.output.precision}")

    variables = Variables(VARIABLES)
    response = Response(RESPONSES, variables)
    interface = study.interface(INTERFACE)
    model = study.model.simulation(MODEL, variables, interface, response)
    sampling = study.method.sampling(METHOD, model)

    print("Running sampling study...")
    study.run(sampling)

    print("Completed DI study.")
    print(f"Samples evaluated: {sampling.num_responses()}")
    if sampling.num_responses() > 0:
        print(f"First response value: {sampling.first_response_value()}")


if __name__ == "__main__":
    main()
